tweets/2022-04-27/2022-04-27-tweets.md

```diff
@@ -0,0 +1,342 @@
+  customElements.define('hello-wave', await import('#components/HelloWave').then(m => m.HelloWave));
+}
+/* END - Rocket auto generated - do not touch */
+
+export default () => html`
+  <h1>Hello World</h1>
+  <hello-wave></hello-wave>
+`;
+```
+
+```html
+<h2>Hello World</h2>
+<hello-wave>
+  <template shadowroot="open">
+    <style>
+      h2 {
+        color: #e03131;
+      }
+    </style>
+    <!--lit-part 18Qb0exw8us=-->
+    <h2>Hello 👋</h2>
+    <!--/lit-part-->
+  </template>
+</hello-wave>
+```
+
+![Hello Wave Static](./wave-static.png)
+
+---
+
+Thats it! 🎉
+
+For components that only require styles and no interactivity we are done.
+
+➡️ The HTML output works without any JavaScript
+➡️ It's crystal clear which components are used on the page
+➡️ You do not need to hand write the imports
+
+so we are done? 🤔
+
+---
+
+Actually - this is where the fun starts 🎉
+
+Let's add a more interactive element
+✍️ an type writer effect
+🔼 with a counter 🤯
+
+```js
+import { LitElement, html, css } from 'lit';
+
+let i = 0;
+const fullText = [...'to this wonderful world of progressive hydration 🤯'];
+
+export class HelloTyper extends LitElement {
+  static properties = {
+    msg: { type: String },
+    counter: { type: Number },
+  };
+
+  constructor() {
+    super();
+    this.msg = ' ';
+    this.counter = 0;
+  }
+
+  updated(changedProperties) {
+    super.updated(changedProperties);
+    if (i < fullText.length) {
+      setTimeout(() => {
+        this.msg += fullText[i];
+        i += 1;
+      }, Math.floor(Math.random() * 50) + 40);
+    }
+  }
+
+  render() {
+    return html`
+      <p>🤔 Hello <span>${this.msg}</span>${'🤯'.repeat(this.counter)}</p>
+      <button @click=${this._inc}>+</button>
+    `;
+  }
+
+  _inc() {
+    if (i >= fullText.length) {
+      this.counter += 1;
+    }
+  }
+
+  static styles = [
+    css`
+      button {
+        font-size: 200%;
+        width: 64px;
+        height: 64px;
+        border: none;
+        border-radius: 10px;
+        background-color: seagreen;
+        color: white;
+      }
+    `,
+  ];
+}
+```
+
+---
+
+Now if we add this to our page we get...
+
+✅ it rendered
+❌ no typing effect
+❌ button is not doing anything
+
+```js
+export async function registerCustomElements() {
+  // server-only components
+  customElements.define('hello-wave', await import('#components/HelloWave').then(m => m.HelloWave));
+  customElements.define('hello-typer', await import('#components/HelloTyper').then(m => m.HelloTyper));
+}
+
+export default () => html`
+  <h1>Hello World</h1>
+  <hello-wave></hello-wave>
+  <hello-typer></hello-typer>
+`;
+```
+
+![Typer Static](./typer-static.png)
+
+---
+
+We get this because our component is now more then "just" text.
+It now has interactive functionality.
+
+Which means that if we only sent HTML & CSS to the browser it will not be enough.
+
+---
+
+We can now decide to skip the server render and render the component purely client side.
+
```

Comment:
Maybe add a sentence here before the bullet points? Feels a bit off right now.

Maybe something like 

"This results in..."

➡️ "nothing" and then the component gets rendered
etc.